### PR TITLE
DDF-1313: Added OSGi commands to create and update a local GeoNames i…

### PR DESCRIPTION
**Pull request by jrnorth**
…ndex

@pklinef @stustison @millerw8 @jlcsmith @kcwire @wmcnalli @beyelerb @emanns95 @coyotesqrl

[Review on Reviewable](https://reviewable.io/reviews/codice/ddf-spatial/39)

### CODE REVIEW

**Comment by coyotesqrl** [comment]
I'm not going to review in depth until Monday morning, but on quick inspection, I see that you didn't add this new functionality to the spatial-app's features.xml file. Do we want it to be automatically installed with spatial, or is it an optional feature?

**Comment by jrnorth** [comment]
I hadn't thought about that. I imagine we'd want it to be installed automatically, but I'm not the most qualified person to ask.

**Comment by jrnorth** [comment]
What sorts of queries are going to be run against this index (i.e., what are we going to be searching for)? The "schema" I've implemented here may need to be adjusted.

Also, indexing a large file (~ 10 million lines) takes a few minutes, so we may want to consider adding some sort of progress indicator.

**Comment by coyotesqrl** [comment]
We have progress indicators in other console commands: ingest comes to mind.

File: spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                    "http://download.geonames.org/export/dump and have the format described " +
+                    "there.", required = true)
+    private String fileLocation = null;
+
+    @Override
+    protected Object doExecute() {
+        final String inputFilesLocation = GeoNamesCreateCommand
+                .getInputFilesLocation(fileLocation);
+
+        if (inputFilesLocation != null) {
```
Comment (coyotesqrl) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java]: I'm not sure why there are both `add` and `create` commands when they both do the same thing.

File: spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesCreateCommand.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    static void buildIndex(final String inputFilesLocation, final boolean create) {
+        CONSOLE.println("Building index...");
+
+        Directory directory;
+
+        try {
+            directory = FSDirectory.open(new File(GEONAMES_INDEX));
+        } catch (IOException e) {
+            CONSOLE.println("Couldn't open the directory for the index.");
+            LOGGER.error("Error opening the directory for the GeoNames index", e);
```
Comment (coyotesqrl) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesCreateCommand.java]: Is there any intervention the user can do at this point to resolve the problem with the index? Should we maybe provide a pointer to what that might be?

**Comment by coyotesqrl** [comment]
:+1:

File: spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java (hunk 1 above)
Comment (jrnorth) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java]: `add` adds entries to an existing index and `create` will create the index (or overwrite an existing one).

Comment (coyotesqrl) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java]: Took me four times through until I saw the `boolean` flag difference. Wow. Rough Monday.

File: spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesCreateCommand.java (hunk 2 above)
Comment (jrnorth) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesCreateCommand.java]: I'm not sure if there is. At this point, there would be an error if somehow the `geonames-index` folder couldn't be opened or created. Perhaps we could include the folder name in the error message that's displayed to the user.

**Comment by wmcnalli** [comment]
:+1:

**Comment by emanns95** [comment]
+1

File: spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.felix.gogo.commands.Command;
+import org.apache.karaf.shell.console.OsgiCommandSupport;
+
+@Command(scope = "geonames", name = "add",
+        description = "Adds to the local GeoNames index, if it exists.")
+public final class GeoNamesAddCommand extends OsgiCommandSupport {
+    @Argument(index = 0, name = "file_location",
+            description = "Location of the GeoNames .txt file or directory containing GeoNames " +
+                    ".txt files.\n" +
+                    "If it's a directory, the contents of all the .txt files inside will be " +
```
Comment (kcwire) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java]: what about just allowing the zip file? I think it would be nice to be able to pull down each of the zip files (allcountries, cities*, etc) and ingest those at once.

File: spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java (hunk 1 above)
Comment (kcwire) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java]: Why not just create a flag for the `add` command to overwrite existing index.  The default could be to simply create & append.  This would get rid of these very similar commands.

File: spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesCreateCommand.java (hunk 2 above)
Comment (kcwire) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesCreateCommand.java]: Can we specify where we want the index to be? I think the default should be in ${DDF_HOME}/data but it wouldn't hurt to allow an admin to specify an external directory either.

Comment (jrnorth) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesCreateCommand.java]: Yeah, that sounds like a good idea.

**Comment by coyotesqrl** [comment]
:+1: On the progress bar.

File: spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java (hunk 1 above)
Comment (jrnorth) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java]: Agreed.

File: spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java (hunk 3 above)
Comment (jrnorth) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesAddCommand.java]: Done.

**Comment by coyotesqrl** [comment]
:+1:

**Comment by jrnorth** [comment]
@kguerriero @wmcnalli @emanns95 @kcwire @millerw8 @jlcsmith @beyelerb @pklinef @stustison 
Even if you've already taken a look at this PR earlier this week, I've made some significant structural changes that warrant another look.

**Comment by kcwire** [comment]
Checkout the travis build - it is failing for some reason.

File: geo/pom.xml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>geo</artifactId>
+    <name>DDF :: Spatial :: Geo</name>
+    <packaging>pom</packaging>
+    <version>2.8.0-SNAPSHOT</version>
+
+    <modules>
+        <module>spatial-geo-api</module>
```
Comment (kcwire) [inline geo/pom.xml:32]: I'm not crazy about `spatial-geo-*` but I don't have another idea off the top of my head.

File: geo/spatial-geo-api/src/main/java/org/codice/ddf/spatial/geo/GeoIndexer.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ **/
+
+package org.codice.ddf.spatial.geo;
+
+public interface GeoIndexer {
```
Comment (kcwire) [inline geo/spatial-geo-api/src/main/java/org/codice/ddf/spatial/geo/GeoIndexer.java:18]: Need javadoc.

**Comment by coyotesqrl** [comment]
I think that Travis build is outside Joe's control. It looks like a problem with a testing maven repo?

Do our multiple Travis builds run in parallel? Is it possible that each of them is stomping on the other in an attempt to push the snapshot?

File: geo/spatial-geo-geonamesindex/src/main/java/org/codice/ddf/spatial/geo/indexer/GeoNamesIndexer.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    @Override
+    public void updateIndex(final String fileLocation, final boolean create) {
+        final String inputFileLocation = getInputFileLocation(fileLocation);
+
+        if (inputFileLocation != null) {
+            buildIndex(inputFileLocation, create);
+        }
+    }
+
+    private String getInputFileLocation(final String fileLocation) {
```
Comment (kcwire) [inline geo/spatial-geo-geonamesindex/src/main/java/org/codice/ddf/spatial/geo/indexer/GeoNamesIndexer.java:63]: Now I'm thinking - why not a URL to a file on a webserver? :) - maybe a future ticket.

File: spatial-commands/pom.xml [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    <artifactId>spatial-commands</artifactId>
+    <name>DDF :: Spatial :: Commands</name>
+    <packaging>bundle</packaging>
+    <version>2.8.0-SNAPSHOT</version>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.codice.ddf.spatial</groupId>
+            <artifactId>spatial-geo-api</artifactId>
+            <version>2.8.0-SNAPSHOT</version>
```
Comment (kcwire) [inline spatial-commands/pom.xml:35]: use ${project.version} so it stays in sync.

File: geo/spatial-geo-geonamesindex/src/main/java/org/codice/ddf/spatial/geo/indexer/GeoNamesIndexer.java (hunk 6 above)
Comment (coyotesqrl) [inline geo/spatial-geo-geonamesindex/src/main/java/org/codice/ddf/spatial/geo/indexer/GeoNamesIndexer.java:63]: Creep. 

Um...I mean _scope creep_.

File: spatial-commands/pom.xml [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.karaf.shell</groupId>
+            <artifactId>org.apache.karaf.shell.console</artifactId>
+            <version>2.4.1</version>
```
Comment (kcwire) [inline spatial-commands/pom.xml:40]: this is probably being pulled in somewhere in dependencyManagement, you may not need to specify the version number here.

File: spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesUpdateCommand.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+@Command(scope = "geonames", name = "update",
+        description = "Creates a local GeoNames index or adds to an existing one.")
+public final class GeoNamesUpdateCommand extends OsgiCommandSupport {
+    @Argument(index = 0, name = "file_location",
+            description = "The absolute path to the GeoNames file whose contents you wish to " +
+                "insert into the index.", required = true)
+    private String inputLocation = null;
+
+    @Option(name = "-c", aliases = "--create",
+            description = "Create a new index or overwrite the existing one.")
```
Comment (kcwire) [inline spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/GeoNamesUpdateCommand.java:32]: So which does this flag do? - Make it very explicit what is the default behavior and what this flag enables.
